share/availability/discovery/metrics.go

```diff
@@ -0,0 +1,166 @@
+	handlePeerConnErr     handlePeerResult = "conn_err"
+	handlePeerInSet       handlePeerResult = "in_set"
+
+	advertiseFailedKey = "failed"
+)
+
+var (
+	meter = global.MeterProvider().Meter("share_discovery")
+)
+
+type handlePeerResult string
+
+type metrics struct {
+	peersAmount      asyncint64.Gauge
+	discoveryResult  syncint64.Counter // attributes: enough_peers[bool],is_canceled[bool]
+	handlePeerResult syncint64.Counter // attributes: result[string]
+	advertise        syncint64.Counter // attributes: failed[bool]
+	peerAdded        syncint64.Counter
+	peerRemoved      syncint64.Counter
+}
+
+// WithMetrics turns on metric collection in discoery.
+func (d *Discovery) WithMetrics() error {
+	metrics, err := initMetrics(d)
+	if err != nil {
+		return fmt.Errorf("discovery: init metrics: %w", err)
+	}
+	d.metrics = metrics
+	d.WithOnPeersUpdate(metrics.observeOnPeersUpdate)
+	return nil
+}
+
+func initMetrics(d *Discovery) (*metrics, error) {
+	peersAmount, err := meter.AsyncInt64().Gauge("discovery_amount_of_peers",
+		instrument.WithDescription("amount of peers in discovery set"))
+	if err != nil {
+		return nil, err
+	}
+
+	findPeersResult, err := meter.SyncInt64().Counter("discovery_find_peers_result",
+		instrument.WithDescription("result of find peers run"))
+	if err != nil {
+		return nil, err
+	}
+
+	handlePeerResultCounter, err := meter.SyncInt64().Counter("discovery_handler_peer_result",
+		instrument.WithDescription("result handling found peer"))
+	if err != nil {
+		return nil, err
+	}
+
+	advertise, err := meter.SyncInt64().Counter("discovery_advertise_event",
+		instrument.WithDescription("advertise events counter"))
+	if err != nil {
+		return nil, err
+	}
+
+	peerAdded, err := meter.SyncInt64().Counter("discovery_add_peer",
+		instrument.WithDescription("add peer to discovery set counter"))
+	if err != nil {
+		return nil, err
+	}
+
+	peerRemoved, err := meter.SyncInt64().Counter("discovery_remove_peer",
+		instrument.WithDescription("remove peer from discovery set counter"))
+	if err != nil {
+		return nil, err
+	}
+
+	metrics := &metrics{
+		peersAmount:      peersAmount,
+		discoveryResult:  findPeersResult,
+		handlePeerResult: handlePeerResultCounter,
+		advertise:        advertise,
+		peerAdded:        peerAdded,
+		peerRemoved:      peerRemoved,
+	}
+
+	err = meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			peersAmount,
+		},
+		func(ctx context.Context) {
+			peersAmount.Observe(ctx, int64(d.set.Size()))
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("registering metrics callback: %w", err)
+	}
+	return metrics, nil
+}
+
+func (m *metrics) observeFindPeers(canceled, isEnoughPeers bool) {
+	if m == nil {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), observeTimeout)
+	defer cancel()
+
+	m.discoveryResult.Add(ctx, 1,
+		attribute.Bool(discoveryFindCancledKey, canceled),
+		attribute.Bool(discoveryEnougPeersKey, isEnoughPeers))
+}
```

Comment:
I saw we pass custom context every time we emot metric, but I don't understand why do this.
* We always have context in the places where we emit metric and it's legit context, which if canceled also cancels metric operations
* The metrics emissions are non-blocking according to the implementation, as it just buffers them and then asynchronously sends them, but if the underlying metrics emitor is blocking this won't respect the caller's context.
* It's just heavy to launch a new context for each internal call everywhere. New context means a new routine at least with timers inside which are not free as well.

So I would like to understand why we follow this pattern, it does not seem right to me.

Reply:
From a sync discussion with @walldiss, we agreed to move without spawning new context just for metrics. The agreed approach is to use the user-given context. In case it is canceled, we will pass context.Background to still track canceled operations in metrics. 

@walldiss, also proposed to make a PR that changes this throughout codebase